src/request/plan.rs

```diff
@@ -63,42 +66,230 @@ impl<Req: KvRequest + HasKeys> HasKeys for Dispatch<Req> {
+            }
+        } else {
+            Ok(vec![Ok(resp)])
+        }
+    }
+
+    // Returns
+    // 1. Ok(true): error has been resolved, retry immediately
+    // 2. Ok(false): backoff, and then retry
+    // 3. Err(Error): can't be resolved, return the error to upper level
+    async fn handle_region_error(
+        pd_client: Arc<PdC>,
+        error: Error,
+        region_store: RegionStore,
+    ) -> Result<bool> {
+        match error {
+            Error::RegionError(mut e) => {
+                let ver_id = region_store.region_with_leader.ver_id();
+                if e.has_not_leader() {
+                    let not_leader = e.get_not_leader();
+                    if not_leader.has_leader() {
+                        match pd_client
+                            .update_leader(
+                                region_store.region_with_leader.ver_id(),
+                                not_leader.get_leader().clone(),
+                            )
+                            .await
+                        {
+                            Ok(_) => Ok(true),
+                            Err(Error::EntryNotFoundInRegionCache) => Ok(false),
+                            Err(e) => {
+                                pd_client.invalidate_region_cache(ver_id).await;
+                                Err(e)
+                            }
+                        }
+                    } else {
+                        // The peer doesn't know who is the current leader. Generally it's because
+                        // the Raft group is in an election, but it's possible that the peer is
+                        // isolated and removed from the Raft group. So it's necessary to reload
+                        // the region from PD.
+                        pd_client.invalidate_region_cache(ver_id).await;
+                        Ok(false)
+                    }
+                } else if e.has_store_not_match() {
+                    pd_client.invalidate_region_cache(ver_id).await;
+                    Ok(false)
+                } else if e.has_epoch_not_match() {
+                    Self::on_region_epoch_not_match(
+                        pd_client.clone(),
+                        region_store,
+                        e.take_epoch_not_match(),
+                    )
+                    .await
+                } else if e.has_server_is_busy()
+                    || e.has_stale_command()
+                    || e.has_region_not_found()
+                    || e.has_max_timestamp_not_synced()
+                {
+                    pd_client.invalidate_region_cache(ver_id).await;
+                    Ok(false)
+                } else if e.has_raft_entry_too_large() {
+                    Err(Error::RegionError(e))
+                } else {
+                    pd_client.invalidate_region_cache(ver_id).await;
+                    Ok(false)
+                }
+            }
+            // errors from PD requests, backoff and retry
```

Comment:
Emm, I don't see why there can be PD errors.

Reply:
I didn't considerably think whether it is possible. Just in case :)

Reply:
The function name is `handle_region_error`, so it is unreasonable that non-region errors occur here. Maybe the definition of `HasRegionError::region_error` can be changed to just return an `errorpb::Error)`.

But it indeed reminds me that the error handling for PD errors is coarse grained. Not a big problem for now as there is typically not much trouble PD can cause.

Reply:
>The function name is handle_region_error, so it is unreasonable that non-region errors occur here.

Right. I just made the change.

Reply:
> Maybe the definition of HasRegionError::region_error can be changed to just return an errorpb::Error)

Good idea